keps/sig-node/5620-resizing-balloon/README.md

```diff
@@ -0,0 +1,812 @@
+
+### Goals
+
+ - Provide autoscalers the ability to upsize and downsize Kubernetes nodes in concert with the mechanisms provided by the underlying cloud provider.
+ - Provide an easy way for users to run multiple Kubelets on the same physical host without overloading the underlying host hardware.
+
+### Non-Goals
+
+ - We do not attempt to provide a way to add underlying hardware in this KEP; this can be done either through hotplug or through mechanisms provided by the cloud provider themselves.
+
+## Proposal
+
+We propose that Kubernetes provide an official "balloon pod" concept. Balloon pods are deployed to each node in the sytem using a daemon set controller which can be enabled or disabled by administrators on a given cluster. Balloon pods claim resources using requests, but then do not consume them. This allows the underlying infrastructure to provide only the resources that are unused by the balloons, in whatever way the infrastructure feels appropriate.
+
+To resize a given node, a scaling component can simply invoke in-place-pod-updates on the resources for the balloon pod. To "upsize" the node the component would "downsize" the balloon pod to release more resources to the pods on the node. To "downsize" the node the component would "upsize" the ballon pod to remove resources from the pool available to the pods on the node.
+
+The only distinction between a balloon pod and a standard pod will be annnotations and priority. Balloon pods should get annotations that can be read by system components and used to distinguish between resources "in-use" by normal pods and resources that "don't exist" (used by balloon pods). In addition, balloon pods will have a special priority which makes them as unpreemptable as system-node-critical pods, but unable to preempt other pods when upsized, to ensure that it does not disrupt running customer pods.
+
+### User Stories (Optional)
+
+#### Autoscaling driven VM resizing
+
+A cloud provider would like to provide dynamically resizable Kubernetes nodes. The cloud provider creates a way to manage the resources provided to a particular Kubelet. By enabling the balloon pods and linking the management of the balloon pod sizes with the underlying resources available to the Kubelet host, the cloud provider can upsize and downsize the Kubernetes nodes to the Kubernetes system without having to involve Kubernetes in the specifics of the cloud resizing mechanism.
+
+#### Multiple Kubelets per node for testing
+
+A user who is testing some Kubernetes feature would like to run many Kubelets on the same host to decrease the amount of resources needed to test scenarios with large numbers of Kubelets. By enabling balloons and then resizing the balloons to ensure each Kubelet only consumes one Nth of the host, the customer can place N Kubelets on the same host without overloading the host itself.
```

Comment:
This usecase seems to be better served by https://github.com/kubernetes/enhancements/pull/5319